File: lexus_scraper.js
const { chromium } = require('playwright');
const fs = require('fs');

async function scrapeLexusDealers() {
  const browser = await chromium.launch({ 
    headless: false, 
    slowMo: 2000,
    args: ['--start-maximized', '--window-size=1400,900']
  });
  const page = await browser.newPage();
  
  // Set viewport to make sure it's visible
  await page.setViewportSize({ width: 1400, height: 900 });
  
  try {
    // Navigate to Lexus dealer page
    await page.goto('https://www.lexus.com/dealers');
    await page.waitForLoadState('networkidle');
    
    // Read existing Lexus.json
    const lexusData = JSON.parse(fs.readFileSync('Lexus.json', 'utf8'));
    
    // States to process (starting from Kentucky)
    const statesToProcess = [
      'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 
      'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
      'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
      'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
      'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina', 'South Dakota',
      'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia',
      'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'
    ];
    
    for (const state of statesToProcess) {
      console.log(`\n🌎 Processing ${state}...`);
      
      // Click on STATE dropdown
      console.log(`   📍 Clicking STATE dropdown...`);
      await page.click('[role="combobox"]');
      await page.waitForTimeout(3000); // Longer delay to see the dropdown open
      
      // Select the state
      console.log(`   🎯 Selecting ${state}...`);
      await page.click(`[role="option"]:has-text("${state}")`);
      await page.waitForLoadState('networkidle');
      await page.waitForTimeout(3000); // Wait to see the results load
      
      // Extract dealer data
      const dealers = await page.evaluate(() => {
        const dealerContainers = document.querySelectorAll('[data-testid="DealerCard"]');
        const dealers = [];
        
        dealerContainers.forEach(container => {
          try {
            const nameElement = container.querySelector('[data-testid="Typography"]');
            const name = nameElement ? nameElement.textContent.trim() : '';
            
            const addressElements = container.querySelectorAll('div');
            let address = '';
            let cityStateZip = '';
            
            for (let i = 0; i < addressElements.length; i++) {
              const text = addressElements[i].textContent.trim();
              if (text && !text.includes('Contact:') && !text.includes('Hours:') && 
                  !text.includes('SE HABLA') && !text.includes('Mon') && 
                  !text.includes('Tue') && !text.includes('Wed') && 
                  !text.includes('Thu') && !text.includes('Fri') && 
                  !text.includes('Sat') && !text.includes('Sun') &&
                  !text.includes('am') && !text.includes('pm') &&
                  !text.includes('Elite') && !text.includes('Monogram') &&
                  !text.includes('Certified') && !text.includes('Collision') &&
                  !text.includes('DEALER') && !text.includes('SITE') &&
                  !text.includes('SERVICE') && !text.includes('CONTACT') &&
                  text !== name) {
                if (text.includes(',')) {
                  cityStateZip = text;
                } else if (text.length > 5) {
                  address = text;
                }
              }
            }
            
            const phoneElement = container.querySelector('a[href^="tel:"]');
            const phone = phoneElement ? phoneElement.textContent.trim() : '';
            
            const hours = [];
            const hoursList = container.querySelector('ul');
            if (hoursList) {
              const hoursItems = hoursList.querySelectorAll('li');
              hoursItems.forEach(item => {
                const dayElement = item.querySelector('div:first-child');
                const timeElement = item.querySelector('div:last-child');
                if (dayElement && timeElement) {
                  hours.push(`${dayElement.textContent.trim()} ${timeElement.textContent.trim()}`);
                }
              });
            }
            
            const badges = [];
            const badgeElements = container.querySelectorAll('img[alt*="Dealer"], img[alt*="Elite"], img[alt*="Monogram"], img[alt*="Certified"]');
            badgeElements.forEach(badge => {
              const altText = badge.getAttribute('alt');
              if (altText && !altText.includes('Tooltip') && !altText.includes('Icon')) {
                badges.push(altText);
              }
            });
            
            const links = {};
            const linkElements = container.querySelectorAll('a');
            linkElements.forEach(link => {
              const href = link.getAttribute('href');
              const text = link.textContent.trim();
              
              if (href && text) {
                if (text.includes('DEALER DETAILS')) {
                  links.dealer_details = href;
                } else if (text.includes('DEALER SITE')) {
                  links.website = href;
                } else if (text.includes('SCHEDULE SERVICE')) {
                  links.service = href;
                } else if (text.includes('CONTACT DEALER')) {
                  links.contact = href;
                }
              }
            });
            
            if (name) {
              dealers.push({
                name: name.toUpperCase(),
                address: address,
                city_state_zip: cityStateZip,
                phone: phone,
                hours: hours,
                badges: badges,
                links: links
              });
            }
          } catch (error) {
            console.error('Error extracting dealer data:', error);
          }
        });
        
        return dealers;
      });
      
      // Add dealers to the data
      if (dealers.length > 0) {
        lexusData.states[state] = dealers;
        lexusData.total_dealers += dealers.length;
        console.log(`   ✅ Added ${dealers.length} dealers for ${state}`);
        console.log(`   📊 Total dealers so far: ${lexusData.total_dealers}`);
      } else {
        lexusData.states[state] = [];
        console.log(`   ⚠️  No dealers found for ${state}`);
      }
      
      // Save updated data
      console.log(`   💾 Saving data to Lexus.json...`);
      fs.writeFileSync('Lexus.json', JSON.stringify(lexusData, null, 2));
      
      // Wait a bit before next state
      console.log(`   ⏳ Waiting before next state...`);
      await page.waitForTimeout(5000); // Longer delay to see the results
    }
    
    console.log('Scraping completed!');
    console.log(`Total dealers: ${lexusData.total_dealers}`);
    
  } catch (error) {
    console.error('Error during scraping:', error);
  } finally {
    await browser.close();
  }
}

scrapeLexusDealers();

File: scripts/lexus_scraper.py
#!/usr/bin/env python3
"""
Lexus Dealer Scraper
Scrapes all Lexus dealers from all 50 US states
"""

import json
import time
from datetime import datetime
from playwright.sync_api import sync_playwright

# List of all 50 US states
STATES = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California", "Colorado",
    "Connecticut", "Delaware", "Florida", "Georgia", "Hawaii", "Idaho",
    "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana",
    "Maine", "Maryland", "Massachusetts", "Michigan", "Minnesota",
    "Mississippi", "Missouri", "Montana", "Nebraska", "Nevada",
    "New Hampshire", "New Jersey", "New Mexico", "New York",
    "North Carolina", "North Dakota", "Ohio", "Oklahoma", "Oregon",
    "Pennsylvania", "Rhode Island", "South Carolina", "South Dakota",
    "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington",
    "West Virginia", "Wisconsin", "Wyoming"
]

def extract_dealer_data(page):
    """Extract dealer data from the current page"""
    dealers = []
    
    # Wait for dealer results to load
    page.wait_for_selector('[data-testid="DealerCard"]', timeout=10000)
    
    # Get all dealer cards
    dealer_cards = page.query_selector_all('[data-testid="DealerCard"]')
    
    for card in dealer_cards:
        try:
            dealer = {}
            
            # Dealer name
            name_elem = card.query_selector('[data-testid="Typography"]')
            if name_elem:
                dealer['name'] = name_elem.inner_text().strip()
            
            # Address
            address_elem = card.query_selector('.sc-bETbji')
            if address_elem:
                address_lines = address_elem.query_selector_all('div')
                if len(address_lines) >= 2:
                    dealer['address'] = address_lines[0].inner_text().strip()
                    dealer['city_state_zip'] = address_lines[1].inner_text().strip()
            
            # Phone number
            phone_elem = card.query_selector('a[href^="tel:"]')
            if phone_elem:
                dealer['phone'] = phone_elem.inner_text().strip()
            
            # Hours
            hours_list = card.query_selector('ul.sc-bsOfxk')
            if hours_list:
                hours = []
                hour_items = hours_list.query_selector_all('li')
                for item in hour_items:
                    day_elem = item.query_selector('div:first-child')
                    time_elem = item.query_selector('div:last-child')
                    if day_elem and time_elem:
                        hours.append(f"{day_elem.inner_text().strip()} {time_elem.inner_text().strip()}")
                dealer['hours'] = hours
            
            # Badges
            badges = []
            
            # Elite Dealer badge
            elite_elem = card.query_selector('[data-testid="Typography"]:has-text("Elite Dealer")')
            if elite_elem:
                badges.append("Elite Dealer")
            
            # Monogram badge
            monogram_elem = card.query_selector('[data-testid="MonogramDealerBadge"]')
            if monogram_elem:
                badges.append("Monogram")
            
            # Collision Center badge
            collision_elem = card.query_selector('[data-testid="CollisionCenterBadge"]')
            if collision_elem:
                badges.append("Certified Collision Center")
            
            dealer['badges'] = badges
            
            # Links
            links = {}
            
            # Dealer details link
            details_elem = card.query_selector('a[href*="/dealers/"]')
            if details_elem:
                links['dealer_details'] = details_elem.get_attribute('href')
            
            # Dealer website
            website_elem = card.query_selector('a[href*="lexus"]:not([href*="/dealers/"])')
            if website_elem:
                links['website'] = website_elem.get_attribute('href')
            
            # Service link
            service_elem = card.query_selector('a:has-text("SCHEDULE SERVICE")')
            if service_elem:
                links['service'] = service_elem.get_attribute('href')
            
            # Contact link
            contact_elem = card.query_selector('a:has-text("CONTACT DEALER")')
            if contact_elem:
                links['contact'] = contact_elem.get_attribute('href')
            
            dealer['links'] = links
            
            dealers.append(dealer)
            
        except Exception as e:
            print(f"Error extracting dealer data: {e}")
            continue
    
    return dealers

def scrape_lexus_dealers():
    """Main function to scrape all Lexus dealers"""
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)
        page = browser.new_page()
        
        try:
            # Navigate to Lexus dealer page
            print("Navigating to Lexus dealer page...")
            page.goto("https://www.lexus.com/dealers")
            
            # Wait for page to load
            page.wait_for_load_state("networkidle")
            
            # Click on STATE tab
            print("Clicking on STATE tab...")
            page.click('[role="tab"]:has-text("STATE")')
            
            # Wait for state dropdown to be available
            page.wait_for_selector('[role="combobox"][aria-label="STATE"]')
            
            all_dealers = {}
            total_dealers = 0
            
            for i, state in enumerate(STATES, 1):
                print(f"Processing {state} ({i}/50)...")
                
                try:
                    # Click on state dropdown
                    page.click('[role="combobox"][aria-label="STATE"]')
                    
                    # Wait for dropdown options
                    page.wait_for_selector('[role="option"]')
                    
                    # Click on the state option
                    page.click(f'[role="option"]:has-text("{state}")')
                    
                    # Wait for results to load
                    page.wait_for_selector('[data-testid="DealerCard"]', timeout=15000)
                    
                    # Extract dealer data
                    dealers = extract_dealer_data(page)
                    
                    if dealers:
                        all_dealers[state] = dealers
                        total_dealers += len(dealers)
                        print(f"  Found {len(dealers)} dealers in {state}")
                    else:
                        print(f"  No dealers found in {state}")
                        all_dealers[state] = []
                    
                    # Small delay between states
                    time.sleep(1)
                    
                except Exception as e:
                    print(f"  Error processing {state}: {e}")
                    all_dealers[state] = []
                    continue
            
            # Create final data structure
            result = {
                "brand": "Lexus",
                "scraped_date": datetime.now().isoformat(),
                "total_dealers": total_dealers,
                "states": all_dealers
            }
            
            # Save to JSON file
            output_file = "/Users/kimseanpardo/autodealerships/Lexus.json"
            with open(output_file, 'w', encoding='utf-8') as f:
                json.dump(result, f, indent=2, ensure_ascii=False)
            
            print(f"\nScraping completed!")
            print(f"Total dealers found: {total_dealers}")
            print(f"Data saved to: {output_file}")
            
        except Exception as e:
            print(f"Error during scraping: {e}")
        
        finally:
            browser.close()

if __name__ == "__main__":
    scrape_lexus_dealers()
